File: src/cs/Naxp.UnitTests/NaxpMessageTests.cs
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Xunit;

namespace LogMu.UnitTests;

/// <summary>
/// The messages themselves: that the table matches the enum, that every member is reachable by
/// name, and that nothing in the wording contradicts how it is used.
/// </summary>
/// <remarks>
/// Three tables have to agree - the enum, the formats beside it, and the rules in
/// <see cref="NaxpMessageRules"/> - and two of them are ordered lists that a member inserted in
/// the middle would silently shift. That is what most of this file is for.
/// </remarks>
public class NaxpMessageTests
{
	static readonly NaxpMessage[] All = (NaxpMessage[])Enum.GetValues(typeof(NaxpMessage));

	[Fact]
	public void Formats_HoldsOneEntryPerMessage()
		=> Assert.Equal(All.Length, NaxpMessages.Count);

	/// <summary>
	/// The number in a member's name is its position, so a member inserted without renumbering is
	/// caught here rather than by a message coming out under the wrong code.
	/// </summary>
	[Fact]
	public void Codes_AreNumberedByPosition()
	{
		for (int i = 0; i < All.Length; ++i)
		{
			string expected = string.Format(CultureInfo.InvariantCulture, "NAXP{0}_", 1001 + i);

			Assert.StartsWith(expected, All[i].ToString(), StringComparison.Ordinal);
		}
	}

	[Fact]
	public void Codes_AreDistinct()
	{
		string[] codes = All.Select(message => message.ToString()).ToArray();

		Assert.Equal(codes.Length, codes.Distinct(StringComparer.Ordinal).Count());
	}

	/// <summary>
	/// Every message says something, and nothing is left as a placeholder.
	/// </summary>
	[Fact]
	public void EveryMessage_HasText()
	{
		foreach (NaxpMessage message in All)
		{
			string text = NaxpMessages.Format(message, null);

			Assert.False(string.IsNullOrWhiteSpace(text), message.ToString());
			Assert.EndsWith(".", text.TrimEnd(), StringComparison.Ordinal);
		}
	}

	/// <summary>
	/// A message that interpolates must be given something, and one that does not must not
	/// silently swallow it. This is what keeps the single argument honest.
	/// </summary>
	[Fact]
	public void MessagesTakingAnArgument_AreExactlyThoseThatUseIt()
	{
		List<NaxpMessage> withArgument = All.Where(TakesAnArgument).ToList();

		Assert.Equal(
			new[]
			{
				NaxpMessage.NAXP1027_RangeReversed,
				NaxpMessage.NAXP1032_EscapeUndefined,
				NaxpMessage.NAXP1033_CharacterNotAllowed,
				NaxpMessage.NAXP1038_ReservedCharacterHere,
				NaxpMessage.NAXP1039_CharacterHere,
				NaxpMessage.NAXP1044_RenderingNotGenerated,
				NaxpMessage.NAXP1046_ReplacementNotSingleValuedWitness,
			},
			withArgument);
	}

	/// <summary>
	/// A message that interpolates must survive being formatted, which a literal brace in it
	/// would prevent.
	/// </summary>
	/// <remarks>
	/// Several messages quote naxps holding braces - <c>'A{2,5}'</c> and <c>Add a '}'</c> among
	/// them - and those never reach <see cref="string.Format(IFormatProvider, string, object?)"/>
	/// because they take no argument. Giving one an argument later without doubling its braces
	/// would throw at the moment of refusal, which is the worst time to find out.
	/// </remarks>
	[Fact]
	public void EveryMessageTakingAnArgument_FormatsWithoutThrowing()
	{
		foreach (NaxpMessage message in All)
		{
			if (!TakesAnArgument(message)) { continue; }

			string text = NaxpMessages.Format(message, "WITNESS");

			Assert.Contains("WITNESS", text, StringComparison.Ordinal);
			Assert.DoesNotContain("{0}", text, StringComparison.Ordinal);
		}
	}

	/// <summary>
	/// The budgets are read from <see cref="NaxpLimits"/> rather than passed in, so a message
	/// naming one must name the figure this build actually ships with.
	/// </summary>
	[Fact]
	public void MessagesNamingABudget_NameTheRealOne()
	{
		string states = NaxpLimits.MaxStates.ToString(CultureInfo.InvariantCulture);

		Assert.Contains(states, NaxpMessages.Format(NaxpMessage.NAXP1049_TooManyStates, null), StringComparison.Ordinal);
		Assert.Contains(states, NaxpMessages.Format(NaxpMessage.NAXP1051_TooManyPairStates, null), StringComparison.Ordinal);

		Assert.Contains(
			NaxpLimits.MaxCanonicalStates.ToString(CultureInfo.InvariantCulture),
			NaxpMessages.Format(NaxpMessage.NAXP1050_TooManyCanonicalStates, null),
			StringComparison.Ordinal);

		Assert.Contains(
			Matcher.MaxGeneratedLength.ToString(CultureInfo.InvariantCulture),
			NaxpMessages.Format(NaxpMessage.NAXP1048_ElementTooLong, null),
			StringComparison.Ordinal);
	}

	[Fact]
	public void EveryMessage_HasARule()
	{
		foreach (NaxpMessage message in All)
		{
			Assert.False(
				string.IsNullOrEmpty(NaxpMessageRules.RuleOf(message)),
				message.ToString());
		}

		Assert.Equal(All.Length, NaxpMessageRules.Mapped.Count);
	}

	/// <summary>
	/// A refusal that names no place in the naxp leaves both numbers at zero, which the public
	/// surface reads as the whole of it. A refusal that does name one must never do that, or it
	/// would be mistaken for the same thing.
	/// </summary>
	[Fact]
	public void ARefusalWithAPosition_IsNotMistakenForTheWholeNaxp()
	{
		Assert.True(new NaxpError(NaxpMessage.NAXP1047_TooManyValues).IsWholeNaxp);
		Assert.False(new NaxpError(NaxpMessage.NAXP1002_IntervalHyphen, offset: 0, length: 1).IsWholeNaxp);
		Assert.False(new NaxpError(NaxpMessage.NAXP1002_IntervalHyphen, offset: 3, length: 1).IsWholeNaxp);
	}

	/// <summary>
	/// The code a caller is given is the number alone, never the hint beside it.
	/// </summary>
	/// <remarks>
	/// A member is spelled <c>NAXP1002_IntervalHyphen</c> so that a line of the library says which
	/// refusal it is about at a glance. That half is a note to ourselves: it would read as a
	/// promise about wording nobody has made, so it stops at the boundary.
	/// </remarks>
	[Fact]
	public void Code_IsTheNumberAloneAndNeverTheHint()
	{
		foreach (NaxpMessage message in All)
		{
			string code = new NaxpError(message).Code;

			Assert.Matches("^NAXP[0-9]{4}$", code);
			Assert.StartsWith(code + "_", message.ToString(), StringComparison.Ordinal);
		}
	}

	/// <summary>
	/// Nor does the hint reach a caller by way of the words, which is the other route out.
	/// </summary>
	[Fact]
	public void NoMessage_QuotesItsOwnMemberName()
	{
		foreach (NaxpMessage message in All)
		{
			var error = new NaxpError(message, TakesAnArgument(message) ? "x" : null);

			Assert.DoesNotContain(message.ToString(), error.Text, StringComparison.Ordinal);
			Assert.DoesNotContain(message.ToString(), error.ToString(), StringComparison.Ordinal);
		}
	}

	/// <summary>
	/// Whether a message interpolates, judged by reading its format rather than by trying it.
	/// </summary>
	static bool TakesAnArgument(NaxpMessage message)
		=> NaxpMessages.Format(message, null).IndexOf("{0}", StringComparison.Ordinal) >= 0;
}

File: src/cpp/tests/naxp_message_tests.cpp
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

#include "check.hpp"
#include "naxp_message_rules.hpp"

#include "fault.hpp"
#include "naxp_limits.hpp"
#include "naxp_message.hpp"

#include <cstddef>
#include <set>
#include <string>
#include <string_view>
#include <vector>

using logmu::detail::fault;
using logmu::detail::format_message;
using logmu::detail::message_code;
using logmu::detail::naxp_message;

namespace
{
	constexpr std::size_t message_count = static_cast<std::size_t>(naxp_message::count);

	naxp_message message_at(std::size_t index) noexcept
	{
		return static_cast<naxp_message>(index);
	}

	bool takes_an_argument(naxp_message message)
	{
		return format_message(message, "").find("{0}") != std::string::npos;
	}
}

NAXP_TEST(codes_are_numbered_by_position)
{
	for (std::size_t i = 0; i < message_count; ++i)
	{
		NAXP_CHECK_EQUAL("NAXP" + std::to_string(1001 + i), std::string(message_code(message_at(i))));
	}
}

NAXP_TEST(codes_are_distinct)
{
	std::set<std::string_view> codes;

	for (std::size_t i = 0; i < message_count; ++i)
	{
		codes.insert(message_code(message_at(i)));
	}

	NAXP_CHECK_EQUAL(message_count, codes.size());
}

NAXP_TEST(every_message_has_text)
{
	for (std::size_t i = 0; i < message_count; ++i)
	{
		const std::string text = format_message(message_at(i), "");

		NAXP_CHECK(!text.empty());
		NAXP_CHECK(text.back() == '.');
	}
}

NAXP_TEST(messages_taking_an_argument_are_exactly_those_that_use_it)
{
	std::vector<naxp_message> with_argument;

	for (std::size_t i = 0; i < message_count; ++i)
	{
		if (takes_an_argument(message_at(i)))
		{
			with_argument.push_back(message_at(i));
		}
	}

	const std::vector<naxp_message> expected = {
		naxp_message::range_reversed,
		naxp_message::escape_undefined,
		naxp_message::character_not_allowed,
		naxp_message::reserved_character_here,
		naxp_message::character_here,
		naxp_message::rendering_not_generated,
		naxp_message::unification_not_single_valued_witness,
		naxp_message::fold_in_character_set,
		naxp_message::repetition_unbounded,
		naxp_message::anchor,
	};

	NAXP_CHECK(expected == with_argument);
}

NAXP_TEST(every_message_taking_an_argument_formats_it_in)
{
	for (std::size_t i = 0; i < message_count; ++i)
	{
		if (!takes_an_argument(message_at(i)))
		{
			continue;
		}

		const std::string text = format_message(message_at(i), "WITNESS");

		NAXP_CHECK(text.find("WITNESS") != std::string::npos);
		NAXP_CHECK(text.find("{0}") == std::string::npos);
	}
}

NAXP_TEST(messages_naming_a_budget_name_the_real_one)
{
	using namespace logmu::detail::limits;

	const std::string states = std::to_string(max_states);

	NAXP_CHECK(format_message(naxp_message::too_many_states, "").find(states) != std::string::npos);
	NAXP_CHECK(format_message(naxp_message::too_many_pair_states, "").find(states) != std::string::npos);
	NAXP_CHECK(format_message(naxp_message::too_many_canonical_states, "").find(std::to_string(max_canonical_states)) != std::string::npos);

	const std::string generated = std::to_string(max_string_length);

	NAXP_CHECK(format_message(naxp_message::element_too_long, "").find(generated) != std::string::npos);
	NAXP_CHECK(format_message(naxp_message::pair_output_abandoned, "").find(generated) != std::string::npos);

	// No message keeps the placeholder the limit is spliced into.
	for (std::size_t i = 0; i < message_count; ++i)
	{
		NAXP_CHECK(format_message(message_at(i), "").find("{limit}") == std::string::npos);
	}
}

NAXP_TEST(every_message_has_a_rule)
{
	for (std::size_t i = 0; i < message_count; ++i)
	{
		NAXP_CHECK(!logmu::testing::rule_of(message_at(i)).empty());
	}
}

NAXP_TEST(a_fault_with_a_position_is_not_mistaken_for_the_whole_naxp)
{
	NAXP_CHECK(fault(naxp_message::too_many_values).is_whole_naxp());
	NAXP_CHECK(!fault(naxp_message::interval_hyphen, 0, 1).is_whole_naxp());
	NAXP_CHECK(!fault(naxp_message::interval_hyphen, 3, 1).is_whole_naxp());
}

NAXP_TEST(a_fault_prints_its_code_and_span_and_text)
{
	NAXP_CHECK_EQUAL(std::string("NAXP1046: ") + format_message(naxp_message::too_many_values, ""), fault(naxp_message::too_many_values).to_string());
	NAXP_CHECK_EQUAL(std::string("NAXP1002 at 3..4: ") + format_message(naxp_message::interval_hyphen, ""), fault(naxp_message::interval_hyphen, 3, 1).to_string());
}
